x/concentrated-liquidity/position.go

```diff
@@ -630,17 +619,46 @@ func (k Keeper) RemovePositionIdToLock(ctx sdk.Context, positionId, underlyingLo
 	store.Delete(lockIdKey)
 }
 
-// positionHasUnderlyingLockInState checks if a given positionId has a corresponding lock in state.
-func (k Keeper) positionHasUnderlyingLockInState(ctx sdk.Context, positionId uint64) (bool, error) {
+// PositionHasActiveUnderlyingLock is a non mutative method that checks if a given positionId has a corresponding lock in state.
+// If it has a lock in state, checks if that lock is still active.
+// If lock is still active, returns true.
+// If lock is no longer active, returns false.
+func (k Keeper) PositionHasActiveUnderlyingLock(ctx sdk.Context, positionId uint64) (hasActiveUnderlyingLock bool, lockId uint64, err error) {
 	// Get the lock ID for the position.
-	_, err := k.GetLockIdFromPositionId(ctx, positionId)
-	if err == nil || !errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
-		return true, nil
-	}
+	lockId, err = k.GetLockIdFromPositionId(ctx, positionId)
 	if errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
-		return false, nil
+		return false, 0, nil
+	} else if err != nil {
+		return false, 0, err
+	}
+
+	// Check if the underlying lock is mature.
+	lockIsMature, err := k.isLockMature(ctx, lockId)
+	if err != nil {
+		return false, 0, err
+	}
+	if lockIsMature {
+		return false, lockId, nil
+	}
+	return true, lockId, nil
+}
+
+// positionHasActiveUnderlyingLockAndUpdate is a mutative method that checks if a given positionId has a corresponding lock in state.
+// If it has a lock in state, checks if that lock is still active.
+// If lock is still active, returns true.
+// If lock is no longer active, removes the lock ID from the position ID to lock ID mapping and returns false.
+func (k Keeper) positionHasActiveUnderlyingLockAndUpdate(ctx sdk.Context, positionId uint64) (hasActiveUnderlyingLock bool, lockId uint64, err error) {
+	hasActiveUnderlyingLock, lockId, err = k.PositionHasActiveUnderlyingLock(ctx, positionId)
+	if err != nil {
+		return false, 0, err
+	}
+	if !hasActiveUnderlyingLock && lockId != 0 {
```

Comment:
Does it make sense to do validation and erorring for other combinations of these conditions?

E.g. `hasActiveUnderlyingLock` but `lockId == 0` should error?

Reply:
Yeah, this should be impossible, but a defense in depth never hurts. Will add